### ISSUE TITLE
how to measure communication costs

### ISSUE THREAD

**Issue by leebs0521** [adap/flower#1827]
### What is your question?

hello, i have a question.

I'm using flwr for federated learning.
i am wondering how to measure commnucation costs between server to clients(not simulations).
Is there any function supported by flwr?

thanks.

**Comment by BaselOmari** [comment]
I'd find this useful for me as well

**Comment by BaselOmari** [comment]
I'll work on implementing it. 

Would you prefer to have a report of the average communication cost across each client per round or the total communication cost per round?

**Comment by BaselOmari** [comment]
@tanertopal @danieljanes would this be something we are open to add? I was thinking of adding it as a metric similar to the loss and accuracy.

**Comment by noobyzy** [comment]
I also have the needs.
I would like to have a `metrics` on the server side as @BaselOmari said, showing the `communication_time` and `computation_time` of each client after each round of `fit`.
And my solution is  as follows: 
1. Measure client training time on the client side, and add a metric called `computation_time` in client's `fit()`
2. Modify the `flwr.server.server.fit_client()`, calculate the execution time of `fit_res = client.fit(ins, timeout=timeout)` , and substract `computation_time` gives `communication_time` . The detailed code is as follows
```py
def fit_client(
        client: ClientProxy, ins: FitIns, timeout: Optional[float]
    ) -> Tuple[ClientProxy, FitRes]:
    """Refine parameters on a single client."""
    start_time = time.time()
    fit_res = client.fit(ins, timeout=timeout)
    total_duration = time.time() - start_time
    fit_res.metrics["total_time"] = total_duration
    fit_res.metrics["communication_time"] = total_duration - float(fit_res.metrics["computation_time"])
    return client, fit_res
```
(one can achieve this by self-define a `flwr.server.Server` class and pass to `flwr.server.start_server()` as an argument)
3. One can retrieve these new metrics using the same way as accessing `loss` and `accuracy`.

I am not sure this method gives accurate results, but according to my tests (I have two types of clients, one using LAN and one using WLAN, and their communication cost ratio is proportional to theirs network speed ratio), I think this works.

**Comment by WilliamLindskog** [comment]
Hi @leebs0521, 

is this an issue that you are still experiencing? If so, you can take a look at https://github.com/adap/flower/tree/main/examples/flower-via-docker-compose or https://github.com/adap/flower/tree/main/examples/advanced-pytorch in which communication is shown.

**Comment by WilliamLindskog** [comment]
Hi again, 

closing this as some of the current example show how this can be done. more info in above comment.
